indexer/SymbolFormatter.cc

```diff
@@ -40,44 +41,194 @@ std::string_view SymbolFormatter::getMacroSymbol(clang::SourceLocation defLoc) {
-      // - generated C++ file: in build tree only (no canonical path), and
-      //   includes header 1 and 2.
-      // We will not emit a symbol that connects the anonymous namespace
-      // in header 1 and header 2. However, that is OK, because it is unclear
-      // how to handle this case anyways, and anonymous namespaces are
-      // rarely (if ever) used in headers.
-      return {};
-    }
-    fmt::format_to(std::back_inserter(this->scratchBuffer), "$ANON/{}\0",
-                   path->asStringView());
-  } else {
-    llvm::raw_string_ostream os(this->scratchBuffer);
-    namespaceDecl->printQualifiedName(os);
-    absl::StrReplaceAll({{"::", "/"}}, &this->scratchBuffer);
+  auto symbol = getSymbol();
+  if (symbol.empty()) {
+    return "";
   }
-  std::string out{
-      fmt::format("c . todo-pkg todo-version {}/", this->scratchBuffer)};
-  this->scratchBuffer.clear();
   auto [newIt, inserted] =
-      this->declBasedCache.insert({namespaceDecl, std::move(out)});
+      this->declBasedCache.insert({decl, std::move(symbol)});
   ENFORCE(inserted);
   return std::string_view(newIt->second);
 }
 
+std::optional<std::string_view>
+SymbolFormatter::getContextSymbol(const clang::DeclContext *declContext) {
+  if (auto namespaceDecl = llvm::dyn_cast<clang::NamespaceDecl>(declContext)) {
+    return this->getNamespaceSymbol(namespaceDecl).value_or("");
+  }
+  if (auto tagDecl = llvm::dyn_cast<clang::TagDecl>(declContext)) {
+    return this->getTagSymbol(tagDecl);
+  }
+  if (llvm::isa<clang::TranslationUnitDecl>(declContext)
+      || llvm::isa<clang::ExternCContextDecl>(declContext)) {
+    auto decl = llvm::dyn_cast<clang::Decl>(declContext);
+    return this->getSymbolCached(
+        decl, [&]() -> std::string { return "c . todo-pkg todo-version "; });
+  }
+  // TODO: Handle all cases of DeclContext here:
+  // Done
+  // - TranslationUnitDecl
+  // - ExternCContext
+  // - NamespaceDecl
+  // - TagDecl
+  // Pending:
+  // - OMPDeclareReductionDecl
+  // - OMPDeclareMapperDecl
+  // - FunctionDecl
+  // - ObjCMethodDecl
+  // - ObjCContainerDecl
+  // - LinkageSpecDecl
+  // - ExportDecl
+  // - BlockDecl
+  // - CapturedDecl
+  return std::nullopt;
+}
+
+std::optional<std::string_view>
+SymbolFormatter::getTagSymbol(const clang::TagDecl *tagDecl) {
+  return this->getSymbolCached(tagDecl, [&]() -> std::string {
+    auto optContextSymbol = this->getContextSymbol(tagDecl->getDeclContext());
+    if (!optContextSymbol.has_value()) {
+      return {};
+    }
+    auto contextSymbol = optContextSymbol.value();
+    if (tagDecl->getDeclName().isEmpty()) {
+      auto definitionTagDecl = tagDecl->getDefinition();
+      ENFORCE(definitionTagDecl, "can't forward-declare an anonymous type");
+      auto defLoc = definitionTagDecl->getLocation();
+      auto defPLoc = this->sourceManager.getPresumedLoc(
+          defLoc.isMacroID() ? this->sourceManager.getSpellingLoc(defLoc)
```

Comment:
TODO: Clean this up now or in follow-up PR. We don't need explicit checks since getXYZLoc already performs the relevant checks